### ISSUE TITLE
 "Path" is incompatible with "PathLike[Unknown]"

### ISSUE THREAD

**Issue by njzjz** [microsoft/pyright#4090]
**Describe the bug**

Sometimes, a `list[Path]` cannot be passed to `list[os.PathLike]`.

**To Reproduce**
Code A
```py
import os
from pathlib import Path

def a(x: list[os.PathLike]):
    return

b=[Path() for _ in range(7)]
a(b)
```

Code B
```py
import os
from pathlib import Path

def a(x: list[os.PathLike]):
    return

a([Path() for _ in range(7)])
```

**Expected behavior**
Both code A and code B have no errors.

**Screenshots or Code**
Code A throws the following errors. Code B has no errors.

```
8:3 - error: Argument of type "list[Path]" cannot be assigned to parameter "x" of type "list[PathLike[Unknown]]" in function "a"
    "list[Path]" is incompatible with "list[PathLike[Unknown]]"
      TypeVar "_T@list" is invariant
        "Path" is incompatible with "PathLike[Unknown]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
Completed in 0.845sec
```

**VS Code extension or command-line**

```sh
pyright 1.py
```

**Additional context**

**Comment by erictraut** [comment]
Yes, this behavior is correct. The type parameter for `list` is invariant, so a value of type `list[Path]` is not compatible with a parameter of type `list[PathLike]`.

Mypy generates the same error in this case, as expected.

For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#generic-types).

**Comment by njzjz** [comment]
Thank you! I realize it will work after I explicitly declare the type of `b`.

```py
import os
from pathlib import Path

def a(x: list[os.PathLike]):
    return

b: list[os.PathLike] = [Path() for _ in range(7)]
a(b)
```

**Comment by erictraut** [comment]
Yes, the reason this (and your second code sample above) works is due to [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types).
